crates/ecs/src/lib.rs

```diff
@@ -806,4 +1138,391 @@ mod tests {
+        let mut world = World::default();
+
+        // add archetype index 0
+        world.add_empty_archetype({
+            let mut archetype = Archetype::default();
+            archetype.add_component_vec::<u64>();
+            archetype.add_component_vec::<u32>();
+
+            archetype
+        });
+
+        // add archetype index 1
+        world.add_empty_archetype({
+            let mut archetype = Archetype::default();
+            archetype.add_component_vec::<u32>();
+
+            archetype
+        });
+
+        let e1 = world.create_new_entity();
+        let e2 = world.create_new_entity();
+        let e3 = world.create_new_entity();
+
+        // e1 and e2 are stored in archetype index 0
+        world.store_entity_in_archetype(e1.id, 0);
+        world.store_entity_in_archetype(e2.id, 0);
+        // e3 is stored in archetype index 1
+        world.store_entity_in_archetype(e3.id, 1);
+
+        // insert some components...
+        world.add_component::<u64>(e1.id, 1);
+        world.add_component::<u32>(e1.id, 2);
+
+        world.add_component::<u64>(e2.id, 3);
+        world.add_component::<u32>(e2.id, 4);
+
+        world.add_component::<u32>(e3.id, 6);
+
+        // Act
+        // Borrow all vecs containing u32 from archetypes have the signature u32
+        let vecs_u32 = world.borrow_component_vecs_with_signature::<u32>(&[TypeId::of::<u32>()]);
+
+        // Assert
+        // Collect values from vecs
+        let result: Vec<u32> = vecs_u32
+            .iter()
+            .flat_map(|x| x.as_ref().unwrap().iter().map(|v| v.unwrap()))
+            .collect();
+        println!("{:?}", result);
+
+        assert_eq!(result, vec![2, 4, 6])
+    }
+
+    #[test]
+    fn querying_with_archetypes() {
+        // Arrange
+        let mut world = World::default();
+
+        // add archetype index 0
+        world.add_empty_archetype({
+            let mut archetype = Archetype::default();
+            archetype.add_component_vec::<u64>();
+            archetype.add_component_vec::<u32>();
+
+            archetype
+        });
+
+        // add archetype index 1
+        world.add_empty_archetype({
+            let mut archetype = Archetype::default();
+            archetype.add_component_vec::<u32>();
+
+            archetype
+        });
+
+        let e1 = world.create_new_entity();
+        let e2 = world.create_new_entity();
+        let e3 = world.create_new_entity();
+
+        // e1 and e2 are stored in archetype index 0
+        world.store_entity_in_archetype(e1.id, 0);
+        world.store_entity_in_archetype(e2.id, 0);
+        // e3 is stored in archetype index 1
+        world.store_entity_in_archetype(e3.id, 1);
+
+        // insert some components...
+        world.add_component::<u64>(e1.id, 1);
+        world.add_component::<u32>(e1.id, 2);
+
+        world.add_component::<u64>(e2.id, 3);
+        world.add_component::<u32>(e2.id, 4);
+
+        world.add_component::<u32>(e3.id, 6);
```

Comment:
since the arrangement of these two tests are the exact same, you should extract that into a common setup-function. This will help us a lot down the line when we refactor the implementation and therefore also will need to refactor the tests.